app/src/main/java/com/example/event_finder_capstone/MainActivity.java

```diff
@@ -1,14 +1,150 @@
+                if (response.isSuccessful() && response.body() != null) {
+                    runOnUiThread(() -> {
+                        try {
+                            JsonObject result = response.body();
+                            eventsList.clear();
+                            eventsList.addAll(convertToList(result));
+                            eventsAdapter.notifyDataSetChanged();
+                        } catch (Exception e) {
+                            Toast.makeText(MainActivity.this, "JSON exception error", Toast.LENGTH_SHORT).show();
+                        }
+                    });
+                } else {
+                    Toast.makeText(MainActivity.this, "Query Failed", Toast.LENGTH_SHORT).show();
+                }
+            }
+
+            @Override
+            public void onFailure(@NonNull Call<JsonObject> call, @NonNull Throwable t) {
+                Toast.makeText(MainActivity.this, "Failed to call API", Toast.LENGTH_SHORT).show();
+            }
+        });
+    }
+
+    private Collection<? extends Event> convertToList(JsonObject result) {
+        List<Event> res = new ArrayList<>();
+        JsonArray events = result.getAsJsonArray("events");
+        for (int i = 0; i < events.size(); i++) {
+            JsonObject temp = (JsonObject) events.get(i);
+            Event event = new Event();
+            populateEventInfo(i, event, temp);
+            res.add(event);
+        }
+        return res;
+    }
+
+    private void populateEventInfo(int  i, Event event, JsonObject temp) {
+        event.setName(temp.get("name").getAsString());
+        event.setDescription(temp.get("description").getAsString());
+        event.setImageUrl(temp.get("image_url").getAsString());
+        event.setTimeStart(temp.get("time_start").getAsString());
+        event.setEventSiteUrl(temp.get("event_site_url").getAsString());
+        checkAndSetEventEndTime(event, temp);
+        checkAndSetEventCost(event, temp);
+        formatAndSetEventLocation(temp, event);
+    }
+
+    private void checkAndSetEventCost(Event event, JsonObject temp) {
+        if (!String.valueOf(temp.get("cost")).equals("null")) {
+            event.setCost("$" + temp.get("cost").getAsString() +"0");
+        } else {
+            event.setCost("N/A");
+        }
+    }
+
+    private void checkAndSetEventEndTime(Event event, JsonObject temp) {
+        if (!String.valueOf(temp.get("time_end")).equals("null")) {
+            event.setTimeEnd(temp.get("time_end").getAsString());
+        } else {
+            event.setTimeEnd(temp.get("time_start").getAsString());
+        }
+    }
+
+    private void formatAndSetEventLocation(JsonObject temp, Event event) {
+        String formattedLocation = temp.get("location").getAsJsonObject().get("display_address").toString();
+        formattedLocation = formattedLocation.replaceAll("[\\[\\]\"\"]", " ");
```

Comment:
❓Question:  Can you give me an example of what the location looked like before this line?  I think we might not be leveraging JSON parsing to its fullest here.

Reply:
Ah, this looks like a [JSONArray](https://stleary.github.io/JSON-java/org/json/JSONArray.html), which is essentially a list of values.  Presumably an event could be happening at multiple locations at once.  We should use JSON parsing instead of Regex to extract the first address.